File: backend/src/ontoforge_server/adapters/neo4j/runtime_store.py
"""Neo4j implementation of the runtime store (instance persistence).

Implements the runtime side of the persistence port (see ``core/ports.py``).
Each method owns its session and delegates to the query functions in
``runtime_queries`` (looked up through the module so unit tests can patch
``ontoforge_server.adapters.neo4j.runtime_queries.<fn>``). Physical naming
(PascalCase labels, UPPER_SNAKE_CASE relationship types, index names) is
derived here from the ontology-level type keys the service passes in.
"""

from typing import Any

from neo4j import AsyncDriver

from ontoforge_server.adapters.neo4j import ddl, oql_compiler, runtime_queries
from ontoforge_server.adapters.neo4j.errors import open_session
from ontoforge_server.adapters.neo4j.filters import (
    build_filter_clauses,
    build_search_clause,
)
from ontoforge_server.core.oql import ValidatedQuery


class Neo4jRuntimeStore:
    def __init__(self, driver: AsyncDriver):
        self._driver = driver

    def _session(self):
        """Session whose driver failures surface as ``StoreError`` (rule 4)."""
        return open_session(self._driver)

    # ------------------------------------------------------------------
    # Schema reading (for the runtime schema cache)
    # ------------------------------------------------------------------

    async def get_full_schema(self, ontology_key: str) -> dict | None:
        async with self._session() as session:
            return await runtime_queries.get_full_schema(session, ontology_key)

    async def get_ai_agent_configs(self, ontology_key: str) -> list[dict]:
        async with self._session() as session:
            return await runtime_queries.get_ai_agent_configs(session, ontology_key)

    async def get_saved_queries(self, ontology_key: str) -> list[dict]:
        async with self._session() as session:
            return await runtime_queries.get_saved_queries(session, ontology_key)

    # ------------------------------------------------------------------
    # Vector-index metadata validation
    # ------------------------------------------------------------------

    def validate_vector_indexed_properties(
        self,
        entity_type_key: str,
        properties: dict[str, Any],
        filter_properties: list[str],
        entity_id: str | None = None,
    ) -> None:
        ddl.validate_vector_indexed_properties(
            entity_type_key, properties, filter_properties, entity_id=entity_id
        )

    # ------------------------------------------------------------------
    # Entity instances
    # ------------------------------------------------------------------

    async def create_entity(
        self,
        entity_type_key: str,
        entity_id: str,
        properties: dict,
        embedding: list[float] | None = None,
    ) -> dict:
        async with self._session() as session:
            return await runtime_queries.create_entity(
                session,
                entity_type_key,
                ddl._to_pascal_case(entity_type_key),
                entity_id,
                properties,
                embedding=embedding,
            )

    async def list_entities(
        self,
        entity_type_key: str,
        property_defs: dict[str, Any],
        filters: dict[str, str],
        search: str | None,
        search_property_keys: list[str],
        sort_field: str,
        order: str,
        limit: int,
        offset: int,
    ) -> tuple[list[dict], int]:
        where_clauses, params = build_filter_clauses(
            filters, property_defs, entity_type_key
        )
        if search and search_property_keys:
            clause, search_params = build_search_clause(search, search_property_keys)
            where_clauses.append(clause)
            params.update(search_params)
        async with self._session() as session:
            return await runtime_queries.list_entities(
                session,
                ddl._to_pascal_case(entity_type_key),
                entity_type_key,
                where_clauses,
                params,
                sort_field,
                order,
                limit,
                offset,
            )

    async def get_entity(self, entity_type_key: str, entity_id: str) -> dict | None:
        async with self._session() as session:
            return await runtime_queries.get_entity(
                session, ddl._to_pascal_case(entity_type_key), entity_id
            )

    async def get_entity_by_id(self, entity_id: str) -> dict | None:
        async with self._session() as session:
            return await runtime_queries.get_entity_by_id(session, entity_id)

    async def get_entities_by_ids(self, entity_ids: list[str]) -> dict[str, dict]:
        async with self._session() as session:
            return await runtime_queries.get_entities_by_ids(session, entity_ids)

    async def update_entity(
        self,
        entity_type_key: str,
        entity_id: str,
        set_properties: dict,
        remove_properties: list[str],
        embedding: list[float] | None = None,
        has_embedding_update: bool = False,
    ) -> dict | None:
        async with self._session() as session:
            return await runtime_queries.update_entity(
                session,
                ddl._to_pascal_case(entity_type_key),
                entity_id,
                set_properties,
                remove_properties,
                embedding=embedding,
                has_embedding_update=has_embedding_update,
            )

    async def delete_entity(self, entity_type_key: str, entity_id: str) -> bool:
        async with self._session() as session:
            return await runtime_queries.delete_entity(
                session, ddl._to_pascal_case(entity_type_key), entity_id
            )

    # ------------------------------------------------------------------
    # Document chunks
    # ------------------------------------------------------------------

    async def get_chunk_embeddings_for_entity_property(
        self, entity_id: str, property_key: str
    ) -> dict[str, list[float]]:
        async with self._session() as session:
            return await runtime_queries.get_chunk_embeddings_for_entity_property(
                session, entity_id, property_key
            )

    async def delete_chunks_for_entity_property(
        self, entity_id: str, property_key: str
    ) -> None:
        async with self._session() as session:
            await runtime_queries.delete_chunks_for_entity_property(
                session, entity_id, property_key
            )

    async def create_document_chunks(
        self,
        entity_id: str,
        entity_type_key: str,
        property_key: str,
        chunks: list[dict],
    ) -> None:
        async with self._session() as session:
            await runtime_queries.create_document_chunks(
                session,
                entity_id,
                ddl.document_virtual_label(entity_type_key, property_key),
                chunks,
            )

    async def search_document_chunks(
        self,
        entity_type_key: str,
        property_key: str,
        query_embedding: list[float],
        limit: int,
    ) -> list[dict]:
        async with self._session() as session:
            return await runtime_queries.search_document_chunks(
                session,
                ddl.document_virtual_label(entity_type_key, property_key),
                ddl.document_index_name(entity_type_key, property_key),
                query_embedding,
                limit,
            )

    # ------------------------------------------------------------------
    # Relation instances
    # ------------------------------------------------------------------

    async def create_relation(
        self,
        relation_type_key: str,
        relation_id: str,
        from_entity_id: str,
        to_entity_id: str,
        properties: dict,
    ) -> dict:
        async with self._session() as session:
            return await runtime_queries.create_relation(
                session,
                relation_type_key,
                relation_type_key.upper(),
                relation_id,
                from_entity_id,
                to_entity_id,
                properties,
            )

    async def list_relations(
        self,
        relation_type_key: str,
        property_defs: dict[str, Any],
        filters: dict[str, str],
        from_entity_id: str | None,
        to_entity_id: str | None,
        sort_field: str,
        order: str,
        limit: int,
        offset: int,
    ) -> tuple[list[dict], int]:
        where_clauses, params = build_filter_clauses(
            filters, property_defs, relation_type_key, node_alias="r"
        )
        if from_entity_id:
            where_clauses.append("from._id = $from_entity_id_filter")
            params["from_entity_id_filter"] = from_entity_id
        if to_entity_id:
            where_clauses.append("to._id = $to_entity_id_filter")
            params["to_entity_id_filter"] = to_entity_id
        async with self._session() as session:
            return await runtime_queries.list_relations(
                session,
                relation_type_key.upper(),
                relation_type_key,
                where_clauses,
                params,
                sort_field,
                order,
                limit,
                offset,
            )

    async def get_relation(
        self, relation_type_key: str, relation_id: str
    ) -> dict | None:
        async with self._session() as session:
            return await runtime_queries.get_relation(
                session, relation_type_key.upper(), relation_id
            )

    async def update_relation(
        self,
        relation_type_key: str,
        relation_id: str,
        set_properties: dict,
        remove_properties: list[str],
    ) -> dict | None:
        async with self._session() as session:
            return await runtime_queries.update_relation(
                session,
                relation_type_key.upper(),
                relation_id,
                set_properties,
                remove_properties,
            )

    async def delete_relation(self, relation_type_key: str, relation_id: str) -> bool:
        async with self._session() as session:
            return await runtime_queries.delete_relation(
                session, relation_type_key.upper(), relation_id
            )

    # ------------------------------------------------------------------
    # Graph traversal
    # ------------------------------------------------------------------

    async def get_neighbors(
        self,
        entity_id: str,
        direction: str,
        relation_type_key: str | None,
        limit: int,
    ) -> list[dict]:
        rel_type_filter = relation_type_key.upper() if relation_type_key else None
        async with self._session() as session:
            return await runtime_queries.get_neighbors(
                session, entity_id, direction, rel_type_filter, limit
            )

    # ------------------------------------------------------------------
    # Semantic search
    # ------------------------------------------------------------------

    async def semantic_search(
        self,
        entity_type_key: str,
        property_defs: dict[str, Any],
        query_embedding: list[float],
        limit: int,
        min_score: float | None,
        filters: dict[str, str] | None = None,
    ) -> list[dict]:
        where_clauses: list[str] = []
        filter_params: dict = {}
        if filters:
            where_clauses, filter_params = build_filter_clauses(
                filters, property_defs, entity_type_key, node_alias="n"
            )
        async with self._session() as session:
            return await runtime_queries.semantic_search(
                session,
                ddl._to_pascal_case(entity_type_key),
                entity_type_key,
                query_embedding,
                limit,
                min_score,
                where_clauses=where_clauses if where_clauses else None,
                filter_params=filter_params if filter_params else None,
            )

    async def semantic_search_all(
        self,
        query_embedding: list[float],
        limit: int,
        min_score: float | None,
    ) -> list[dict]:
        """Search the shared cross-type entity vector index."""
        async with self._session() as session:
            return await runtime_queries.semantic_search(
                session,
                "_Entity",
                "",
                query_embedding,
                limit,
                min_score,
                index_name=ddl.ENTITY_VECTOR_INDEX_NAME,
            )

    async def search_saved_queries(
        self,
        query_embedding: list[float],
        ontology_key: str,
        limit: int,
        min_score: float | None,
    ) -> list[dict]:
        async with self._session() as session:
            return await runtime_queries.search_saved_queries(
                session, query_embedding, ontology_key, limit, min_score
            )

    # ------------------------------------------------------------------
    # Compiled read queries (OQL execution)
    # ------------------------------------------------------------------

    async def execute_oql(
        self,
        validated: "ValidatedQuery",
        params: dict[str, Any] | None = None,
    ) -> tuple[list[str], list[dict]]:
        """Compile a validated OQL query to Cypher and execute it read-only."""
        cypher = oql_compiler.compile_query(validated)
        async with self._session() as session:
            return await runtime_queries.execute_cypher_read(
                session, cypher, params=params
            )

File: server/src/adapters/neo4j/runtimeStore.ts
import type { KeywordPropertySegment } from "../../core/ports.js";
/**
 * Neo4j implementation of the runtime store (instance-data persistence).
 *
 * Implements the runtime side of the persistence port (see
 * `core/ports.ts`). Each method owns its session — opened through
 * `runSession`, so driver failures surface as `StoreError` (rule 4) — and
 * delegates to the query functions in `runtimeQueries.ts`. Physical naming
 * (PascalCase labels, UPPER_SNAKE_CASE relationship types) is derived here
 * from the lens-level type keys the service passes in.
 *
 * Write values cross the port in their port-safe forms (JS `Date` for
 * datetimes, ISO strings for dates, plain numbers for integers) and are
 * converted to driver-native types here, guided by the property
 * definitions the service supplies — the driver would otherwise store
 * every number as a float and every temporal as a string.
 *
 * The property definitions the read methods carry are the row-decoding
 * aid for adapters whose storage does not distinguish temporals from
 * text. Neo4j stores native temporals and reads them back as such, so
 * this adapter ignores those parameters.
 */

import type { TextSearchLanguage } from "../../registry/schemas.js";

import neo4j, { type Driver } from "neo4j-driver";

import type { ValidatedQuery } from "../../core/oql/index.js";
import type {
  FilterCondition,
  Row,
  RuntimeStore,
  SearchedType,
  SearchedProperty,
} from "../../core/ports.js";
import type { PropertyDef } from "../../core/schemas.js";
import {
  documentIndexName,
  documentVirtualLabel,
  toPascalCase,
  toUpperSnakeCase,
  validateVectorIndexedProperties,
} from "./ddl.js";
import { runSession } from "./errors.js";
import { buildFilterClauses, buildSearchClause, toNeo4jParameter } from "./filters.js";
import { compileQuery } from "./oqlCompiler.js";
import * as queries from "./runtimeQueries.js";

/** Convert a property map to driver-native parameter values. Internal
 * `_doc_*_length` counters are integers; everything else follows its
 * property definition's data type. */
function toWriteProperties(properties: Row, propertyDefs: Record<string, PropertyDef>): Row {
  const converted: Row = {};
  for (const [key, value] of Object.entries(properties)) {
    if (key.startsWith("_doc_") && key.endsWith("_length")) {
      converted[key] = neo4j.int(value as number);
      continue;
    }
    const def = propertyDefs[key];
    converted[key] = def === undefined ? value : toNeo4jParameter(value, def.dataType);
  }
  return converted;
}

/** The WHERE fragments and bound parameters for a search's filter
 * conditions on the node aliased `alias` — `null` for both when there
 * are none, which is how the query functions read "unfiltered". */
function filterFragments(
  filters: FilterCondition[] | null,
  alias: string,
): [string[] | null, Row | null] {
  if (filters === null || filters.length === 0) {
    return [null, null];
  }
  const [whereClauses, params] = buildFilterClauses(filters, alias);
  return [whereClauses, params];
}

export class Neo4jRuntimeStore implements RuntimeStore {
  /** Bound to one ontology; unbound (tests only) carries the empty key. */
  constructor(
    private readonly driver: Driver,
    public readonly ontologyKey: string = "",
    public readonly textSearchLanguage: TextSearchLanguage = "english",
  ) {}

  // ------------------------------------------------------------------
  // Declarations
  // ------------------------------------------------------------------

  /** Declared unsupported: the in-index WHERE of a vector search cannot
   * express a pattern predicate, so a path condition on search
   * is rejected above the port and the entity list is the alternative
   * (`docs/storage-adapters.md`, the divergence list). */
  supportsKeywordRanking(): boolean {
    return false;
  }

  supportsSearchPathConditions(): boolean {
    return false;
  }

  // ------------------------------------------------------------------
  // Schema reading (for the runtime schema cache)
  // ------------------------------------------------------------------

  async getFullSchema(lensKey: string): Promise<Row | null> {
    return runSession(this.driver, (session) => queries.getFullSchema(session, lensKey));
  }

  async getAiAgentConfigs(lensKey: string): Promise<Row[]> {
    return runSession(this.driver, (session) => queries.getAiAgentConfigs(session, lensKey));
  }

  async getSavedQueries(lensKey: string): Promise<Row[]> {
    return runSession(this.driver, (session) => queries.getSavedQueries(session, lensKey));
  }

  // ------------------------------------------------------------------
  // Vector-index metadata validation
  // ------------------------------------------------------------------

  /** Reject string values too large for vector-index filter metadata.
   * Synchronous; raises the domain `ValidationError` (see `ddl.ts`). */
  validateVectorIndexedProperties(
    entityTypeKey: string,
    properties: Row,
    filterProperties: string[],
    entityId: string | null = null,
  ): void {
    validateVectorIndexedProperties(entityTypeKey, properties, filterProperties, entityId);
  }

  // ------------------------------------------------------------------
  // Entity instances
  // ------------------------------------------------------------------

  async createEntity(
    entityTypeKey: string,
    entityId: string,
    properties: Row,
    propertyDefs: Record<string, PropertyDef>,
    embedding: number[] | null = null,
    _propertyText = "",
    _keywordSegments?: KeywordPropertySegment[],
  ): Promise<Row> {
    return runSession(this.driver, (session) =>
      queries.createEntity(
        session,
        entityTypeKey,
        toPascalCase(entityTypeKey),
        entityId,
        toWriteProperties(properties, propertyDefs),
        embedding,
      ),
    );
  }

  async listEntities(
    entityTypeKey: string,
    _propertyDefs: Record<string, PropertyDef>,
    filters: FilterCondition[],
    search: string | null,
    searchPropertyKeys: string[],
    sortField: string,
    order: string,
    limit: number,
    offset: number,
  ): Promise<[Row[], number]> {
    const [whereClauses, params] = buildFilterClauses(filters);
    if (search !== null && search !== undefined && searchPropertyKeys.length > 0) {
      const [clause, searchParams] = buildSearchClause(search, searchPropertyKeys);
      whereClauses.push(clause);
      Object.assign(params, searchParams);
    }
    return runSession(this.driver, (session) =>
      queries.listEntities(
        session,
        toPascalCase(entityTypeKey),
        entityTypeKey,
        whereClauses,
        params,
        sortField,
        order,
        limit,
        offset,
      ),
    );
  }

  async getEntity(entityTypeKey: string, entityId: string): Promise<Row | null> {
    return runSession(this.driver, (session) =>
      queries.getEntity(session, toPascalCase(entityTypeKey), entityId),
    );
  }

  async getEntityById(
    entityId: string,
    _propertyDefs: Record<string, PropertyDef>,
  ): Promise<Row | null> {
    return runSession(this.driver, (session) => queries.getEntityById(session, entityId));
  }

  async updateEntity(
    entityTypeKey: string,
    entityId: string,
    setProperties: Row,
    removeProperties: string[],
    propertyDefs: Record<string, PropertyDef>,
    embedding: number[] | null = null,
    hasEmbeddingUpdate = false,
    _propertyText = "",
    _keywordSegments?: KeywordPropertySegment[],
  ): Promise<Row | null> {
    return runSession(this.driver, (session) =>
      queries.updateEntity(
        session,
        toPascalCase(entityTypeKey),
        entityId,
        toWriteProperties(setProperties, propertyDefs),
        removeProperties,
        embedding,
        hasEmbeddingUpdate,
      ),
    );
  }

  async deleteEntity(entityTypeKey: string, entityId: string): Promise<boolean> {
    return runSession(this.driver, (session) =>
      queries.deleteEntity(session, toPascalCase(entityTypeKey), entityId),
    );
  }

  // ------------------------------------------------------------------
  // Document chunks
  // ------------------------------------------------------------------

  async getChunkEmbeddingsForEntityProperty(
    entityId: string,
    propertyKey: string,
  ): Promise<Record<string, number[]>> {
    return runSession(this.driver, (session) =>
      queries.getChunkEmbeddingsForEntityProperty(session, entityId, propertyKey),
    );
  }

  async deleteChunksForEntityProperty(entityId: string, propertyKey: string): Promise<void> {
    return runSession(this.driver, (session) =>
      queries.deleteChunksForEntityProperty(session, entityId, propertyKey),
    );
  }

  async createDocumentChunks(
    entityId: string,
    entityTypeKey: string,
    propertyKey: string,
    chunks: Row[],
  ): Promise<void> {
    // Ordinals and coordinates are integers; the driver would otherwise
    // store plain JS numbers as floats.
    const rows = chunks.map((chunk) => ({
      ...chunk,
      _index: neo4j.int(chunk._index as number),
      startChar: neo4j.int(chunk.startChar as number),
      charLength: neo4j.int(chunk.charLength as number),
    }));
    return runSession(this.driver, (session) =>
      queries.createDocumentChunks(
        session,
        entityId,
        documentVirtualLabel(entityTypeKey, propertyKey),
        rows,
      ),
    );
  }

  /** Conditions are built against the parent node alias `n` and applied
   * below the port, after the index lookup (`runtimeQueries.ts`). Only
   * plain conditions arrive: this adapter declares no path-condition
   * support for semantic search, so the service rejects paths above the
   * port before any search runs. */
  async documentSearchSemantic(
    properties: SearchedProperty[],
    embedding: number[],
    limit: number,
  ): Promise<Row[]> {
    return this.rankedSemantic(properties, embedding, limit, true);
  }

  async getEntitiesByIds(
    entityIds: string[],
    _propertyDefs: Record<string, PropertyDef>,
  ): Promise<Record<string, Row>> {
    return runSession(this.driver, (session) => queries.getEntitiesByIds(session, entityIds));
  }

  // ------------------------------------------------------------------
  // Semantic search
  // ------------------------------------------------------------------

  async propertySearchKeyword(): Promise<Row[]> {
    throw new Error("Keyword ranking is not supported");
  }
  async documentSearchKeyword(): Promise<Row[]> {
    throw new Error("Keyword ranking is not supported");
  }

  async propertySearchSemantic(
    types: SearchedType[],
    embedding: number[],
    limit: number,
  ): Promise<Row[]> {
    return this.rankedSemantic(types, embedding, limit, false);
  }

  private async rankedSemantic(
    searched: (SearchedType | SearchedProperty)[],
    embedding: number[],
    limit: number,
    document: boolean,
  ): Promise<Row[]> {
    if (!searched.length) return [];
    const params: Row = { embedding, limit: neo4j.int(limit) };
    const scans = searched.map((item, i) => {
      const [clauses, filterParams] = filterFragments(item.conditions, "n");
      let where = (clauses ?? []).join(" AND ");
      for (const [key, value] of Object.entries(filterParams ?? {}).sort(
        ([a], [b]) => b.length - a.length,
      )) {
        where = where.replaceAll(`$${key}`, `$s${i}_${key}`);
        params[`s${i}_${key}`] = value;
      }
      const label =
        "propertyKey" in item
          ? documentVirtualLabel(item.entityTypeKey, item.propertyKey)
          : toPascalCase(item.entityTypeKey);
      const index =
        "propertyKey" in item
          ? documentIndexName(item.entityTypeKey, item.propertyKey)
          : `${item.entityTypeKey}_embedding`;
      const alias = document ? "c" : "n";
      const parent = document && where ? "MATCH (n:_Entity)-[:_HAS_CHUNK]->(c) " : "WITH n, score ";
      return (
        `MATCH (${alias}:${label}) SEARCH ${alias} IN (VECTOR INDEX ${index} FOR $embedding LIMIT $limit) SCORE AS score ` +
        (where ? `${parent}WHERE ${where} ` : "") +
        `RETURN ${alias} {.*} AS value, score`
      );
    });
    return runSession(this.driver, async (session) => {
      const result = await session.run(
        `CALL () { ${scans.join(" UNION ALL ")} } RETURN value, score ORDER BY score DESC LIMIT $limit`,
        params,
      );
      return result.records.map((record) => ({
        [document ? "chunk" : "entity"]: queries.toEntityRow(record.get("value")),
        score: record.get("score"),
      }));
    });
  }

  /** Rank SavedQuery descriptions for one lens by vector similarity. */
  async searchSavedQueries(
    queryEmbedding: number[],
    lensKey: string,
    limit: number,
    minScore: number | null,
  ): Promise<Row[]> {
    return runSession(this.driver, (session) =>
      queries.searchSavedQueries(session, queryEmbedding, lensKey, limit, minScore),
    );
  }

  // ------------------------------------------------------------------
  // Relation instances
  // ------------------------------------------------------------------

  async createRelation(
    relationTypeKey: string,
    relationId: string,
    fromEntityId: string,
    toEntityId: string,
    properties: Row,
    propertyDefs: Record<string, PropertyDef>,
  ): Promise<Row> {
    return runSession(this.driver, (session) =>
      queries.createRelation(
        session,
        relationTypeKey,
        toUpperSnakeCase(relationTypeKey),
        relationId,
        fromEntityId,
        toEntityId,
        toWriteProperties(properties, propertyDefs),
      ),
    );
  }

  async listRelations(
    relationTypeKey: string,
    _propertyDefs: Record<string, PropertyDef>,
    filters: FilterCondition[],
    fromEntityId: string | null,
    toEntityId: string | null,
    sortField: string,
    order: string,
    limit: number,
    offset: number,
  ): Promise<[Row[], number]> {
    const [whereClauses, params] = buildFilterClauses(filters, "r");
    if (fromEntityId) {
      whereClauses.push("from._id = $from_entity_id_filter");
      params.from_entity_id_filter = fromEntityId;
    }
    if (toEntityId) {
      whereClauses.push("to._id = $to_entity_id_filter");
      params.to_entity_id_filter = toEntityId;
    }
    return runSession(this.driver, (session) =>
      queries.listRelations(
        session,
        toUpperSnakeCase(relationTypeKey),
        relationTypeKey,
        whereClauses,
        params,
        sortField,
        order,
        limit,
        offset,
      ),
    );
  }

  async getRelation(relationTypeKey: string, relationId: string): Promise<Row | null> {
    return runSession(this.driver, (session) =>
      queries.getRelation(session, toUpperSnakeCase(relationTypeKey), relationId),
    );
  }

  async updateRelation(
    relationTypeKey: string,
    relationId: string,
    setProperties: Row,
    removeProperties: string[],
    propertyDefs: Record<string, PropertyDef>,
  ): Promise<Row | null> {
    return runSession(this.driver, (session) =>
      queries.updateRelation(
        session,
        toUpperSnakeCase(relationTypeKey),
        relationId,
        toWriteProperties(setProperties, propertyDefs),
        removeProperties,
      ),
    );
  }

  async deleteRelation(relationTypeKey: string, relationId: string): Promise<boolean> {
    return runSession(this.driver, (session) =>
      queries.deleteRelation(session, toUpperSnakeCase(relationTypeKey), relationId),
    );
  }

  // ------------------------------------------------------------------
  // OQL
  // ------------------------------------------------------------------

  /**
   * Compile a validated OQL query to Cypher and execute it read-only.
   * The validated query crosses the port opaque (`core/ports.ts` rule 1);
   * parameters arrive separately as a map (empty for ad-hoc queries —
   * binding is a saved-query concern). Parameters the analysis recorded
   * as SKIP/LIMIT operands are wrapped as driver integers — a plain JS
   * number crosses the wire as a Float, which the server rejects as a
   * paging count.
   */
  async executeOql(validated: ValidatedQuery, params: Row = {}): Promise<[string[], Row[]]> {
    const cypher = compileQuery(validated);
    const converted: Row = { ...params };
    for (const name of validated.analysis.skipLimitParams) {
      if (name in converted) {
        converted[name] = neo4j.int(converted[name] as number);
      }
    }
    return runSession(this.driver, (session) =>
      queries.executeCypherRead(session, cypher, converted),
    );
  }

  // ------------------------------------------------------------------
  // Graph traversal
  // ------------------------------------------------------------------

  async getNeighbors(
    entityId: string,
    direction: string,
    relationTypeKey: string | null,
    limit: number,
    _propertyDefsByType: Record<string, Record<string, PropertyDef>>,
  ): Promise<Row[]> {
    const relTypeFilter = relationTypeKey ? toUpperSnakeCase(relationTypeKey) : null;
    return runSession(this.driver, (session) =>
      queries.getNeighbors(session, entityId, direction, relTypeFilter, limit),
    );
  }
}
